smart-py/marketplace.py

```diff
@@ -1,64 +1,203 @@
+
+        self.data.swaps[self.data.counter] = sp.record(
+            issuer=sp.sender,
+            objkt_amount=params.objkt_amount,
+            objkt_id=params.objkt_id,
+            xtz_per_objkt=params.xtz_per_objkt,
+            royalties=params.royalties,
+            creator=params.creator
+        )
+
         self.data.counter += 1
-    
+
     @sp.entry_point
     def collect(self, params):
         sp.verify(
             # verifies if tez amount is equal to price per objkt
-            (sp.amount == sp.utils.nat_to_mutez(sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some()))) & (self.data.swaps[params.swap_id].objkt_amount != 0))
+            (
+                sp.amount == sp.utils.nat_to_mutez(
+                    sp.fst(
+                        sp.ediv(
+                            self.data.swaps[params.swap_id].xtz_per_objkt,
+                            sp.mutez(1)
+                        ).open_some()
+                    )
+                )
+            # and verify there are objkts available for swap
+            ) & (
+                self.data.swaps[params.swap_id].objkt_amount > 0
+            # and not collecting own swap
+            ) & (
+                sp.sender != self.data.swaps[params.swap_id].issuer
+            )
+        )
 
         sp.if (self.data.swaps[params.swap_id].xtz_per_objkt != sp.tez(0)):
+            self.amount = sp.fst(
+                sp.ediv(
+                    self.data.swaps[params.swap_id].xtz_per_objkt,
+                    sp.mutez(1)
+                ).open_some()
+            )
 
-            self.amount = sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some())
-                
             # calculate fees and royalties
             self.fee = self.amount * (self.data.swaps[params.swap_id].royalties + self.data.fee) / 1000
+
             self.royalties = self.data.swaps[params.swap_id].royalties * self.fee / (self.data.swaps[params.swap_id].royalties + self.data.fee)
-            
+
             # send royalties to NFT creator
-            sp.send(self.data.swaps[params.swap_id].creator, sp.utils.nat_to_mutez(self.royalties))
-                
+            sp.send(
+                self.data.swaps[params.swap_id].creator,
+                sp.utils.nat_to_mutez(self.royalties)
+            )
+
             # send management fees
-            sp.send(self.data.manager, sp.utils.nat_to_mutez(abs(self.fee - self.royalties)))
-                
+            sp.send(
+                self.data.manager,
+                sp.utils.nat_to_mutez(
+                    abs(
+                        self.fee - self.royalties
+                    )
+                )
+            )
+
             # send value to issuer
```

Comment:
"value" might be a bit overloaded. Would something like this be better: send amount minus fees to issuer

Reply:
👍